### PR TITLE
notes: implement basic bounded search

### PR DESCRIPTION
**Pull request by Fang-**
## Summary

Similar to the search in channels agent introduced in #3293, we add a search endpoint to notes which bounds its results by posts _searched_ rather than posts _returned_. Adds this both as a scry endpoint and an HTTP one.  
Closes TLON-6237.

## Changes

Nothing crazy different from what already exists in channels: optional `from` to specify a location to start the backwards search, `tries` to bound the search and encoded `nedl` for the search string.

We tweak `+find` to do the `+cass` on the _entire_ `.hay` ahead of time, which is more performant in the "search through big content string" scenarios we hit in notebooks.

I stuck with local convention best I could, but couple of things to note about that for explicit review/discussion @arthyn:
- Something's deranged about URL parsing/handling in `+serve-http`: for `/notes/~/v1/request/0vabc.defgh` we take the `.defgh` out of `ext` and put it back onto the request path. We don't do this for any other endpoints. (And so I don't do it here, even though the `@t` encoding needs it.) Is there a reason we don't require `.json` extension everywhere? Or (sorta) require absence of `ext`?  (Except the file serving places obviously, but presence/absence of `/~` infix helps differentiate.)
- Local convention is to emit note IDs as json numbers, but to parse them as `@ud`s in endpoint paths. As a result, you can't just chuck `last` into the next search page URL, you need to put the dots in by hand. LLM will complain about this, but it matches existing behavior.

## How did I test?

Wrote and ran tests for normal cases and all edge cases encountered during development. No client integration yet.

## Risks and impact

- Yes, safe to rollback without consulting PR author.

## Rollback plan

Revert.

### CODE REVIEW

File: desk/app/notes.hoon [hunk 1]
```diff
@@ -2748,6 +2759,79 @@
...
+    ::
+        [%search %bounded %text from=@ tries=@ nedl=@ ~]
+      =*  vars  t.t.t.rest
+      :-  ~
+      %-  scam:enjs:notes-json
+      %^    no-search
+          ?:  =(%$ from.i.vars)  ~
+          `(slav %ud from.i.vars)
+        (slav %ud tries.i.t.vars)
+      (slav %t nedl.i.t.t.vars)
```
[the remark this answers is not in the document]
Comment (Fang-) [reply desk/app/notes.hoon]: Yes, we know, it's in the PR description.

File: desk/app/notes.hoon [hunk 2]
```diff
@@ -2748,6 +2759,79 @@
...
+  ::  +no-search: backwards bounded search
+  ::
+  ++  no-search
+    |=  [from=(unit @ud) tries=@ud nedl=@t]
+    ^-  scam:n
+    =/  notes=(list [nid=@ud =note:n])
+      ::NOTE  filtering for u.from prior to sort actually slightly slower
+      ::      (in tested cases)
+      %+  sort  ~(tap by notes.notebook-state)
+      |=([[a=@ud *] [b=@ud *]] (gth a b))
```
[the remark this answers is not in the document]
Comment (Fang-) [reply desk/app/notes.hoon:2813]: I'm aware. Initial testing showed that cost here (and of walking to the `from`) dwarfs in comparison to running the search operation itself.

Also filtering out entries above `from` _before the sort_ requires a full map/list traversal, so doing that naively all the time will hit cases where the cost exceeds the gain, it's only worth it if `from` points to a much-older-than-latest note, which is not the common case.

File: desk/app/notes.hoon [hunk 3]
```diff
@@ -2748,6 +2759,79 @@
...
+    =/  from=@ud
+      (fall from ?~(notes 0 +(nid.i.notes)))
+    =|  found=(list note:n)
+    =*  done  [from (flop found)]
+    ?:  =(0 from)       done
+    |^  ?~  notes       =.(from 0 done)
+        ?:  =(0 tries)  done  ::  compute out of bounds
```
[the remark this answers is not in the document]
Comment (Fang-) [reply desk/app/notes.hoon:2820]: `tries=0` is bogus, this case doesn't matter.

**Review by mikolajpp — APPROVED** [review]
This looks great. Search and you shall find. Approved!

File: desk/tests/app/notes.hoon [hunk 4]
```diff
@@ -2553,4 +2553,260 @@
...
+  =/  s0=state-15:n  !<(state-15:n sv)
+  =/  key=@t  ?~(api-key.s0 '' u.api-key.s0)
+  =/  hdr=(list [@t @t])  ~[['x-api-key' key]]
+  =/  base=tape
+    "/notes/~/v1/notebooks/{<`@p`ship.f>}/{(trip name.f)}"
+  =/  tries=tape  (trip (scot %ud 5))
+  =/  nedl=tape   (trip (scot %t 'needle'))
+  =/  url=@t
+    (crip "{base}/search/bounded/text//{tries}/{nedl}")
+  (ex-get-200 hdr url)
```
Comment (arthyn) [inline desk/tests/app/notes.hoon]: maybe good to also assert on the response body itself here

File: desk/app/notes.hoon [hunk 5]
```diff
@@ -2692,6 +2693,16 @@
         |=  [who=ship r=role:n]
         [who r]
       ``notes-members+!>(mrecords)
+    ::
+        %search
+      ?.  ?=([%bounded %text from=@ tries=@ nedl=@ ~] rest)
```
Comment (arthyn) [inline desk/app/notes.hoon:2719]: I would use a query param here instead of path part for the nedl so we can avoid the dot shenanigans

**Review by arthyn — CHANGES_REQUESTED** [review]
mostly lgtm, the query param I think is worth doing now

File: desk/app/notes.hoon (hunk 5 above)
Comment (arthyn) [reply desk/app/notes.hoon:2719]: Done in 431d8cf0b: `needle`, `from` and `tries` are all query args now (`.../search/bounded/text?needle=v1.0&from=4&tries=50`). Moved `from`/`tries` too rather than just the needle — as the new last path segment, `tries` would have inherited the same ext-truncation. Both are optional now (no `from` = newest, no `tries` = default budget of 100), and the @ud parse takes a plain `1234` as well as `1.234` so a client can echo `last` back verbatim. Scry path unchanged.

File: desk/app/notes.hoon (hunk 1 above)
[the remark this answers is not in the document]
Comment (arthyn) [reply desk/app/notes.hoon]: Fixed in 431d8cf0b — the needle is a query arg now, so it never passes through apat and dotted needles search the full string (covered by test-search-v1-http-needle-is-plain-text).
